### PR TITLE
Add container query support

### PR DESCRIPTION
**Pull request by MorganWheel**
Supports the use of contain, container, container-name, and container-type for identifying containers. If no container is identified, window dimensions are used instead.

### CODE REVIEW

File: src/make-styles.tsx [hunk 1]
```diff
@@ -203,15 +215,31 @@ export const makeTemplateFunction = <
...
+      
+      let styleProps: { [key: string]: Style | Style[] } = useStyleSheet(styles, theme, dimensions, containerDimensions);
       styleProps = style ? { ...styleProps, style: [styleProps.style, style] } : styleProps;
       const transformedProps = transformProps({ ...props, theme } as AttrProps<P, I, A>);
+
+      if(isContainer) {
+        return (
+          <ContainerSizeContext.Provider value = {componentDimensions[0]}>
+            <Component {...filterComponentProps(transformedProps)} {...styleProps} ref={ref} onLayout={componentDimensions[1]}>
+              {children}
```
Comment (MrLoh) [inline src/make-styles.tsx:238]: I **think** the `ContainerSizeContext.Provider` needs to go on the inside of the component not on the outside. Maybe it doesn't matter just wondering what happens if the container is using a container query itself.

File: src/window-dimensions.ts [hunk 2]
```diff
@@ -1,19 +1,32 @@
-import { useEffect, useState } from 'react';
+import { useEffect, useMemo, useState } from 'react';
 import { Dimensions } from 'react-native';
 
 import type { ScaledSize } from 'react-native';
+import { ComponentSizeChanges } from 'make-styles';
```
Comment (MrLoh) [inline src/window-dimensions.ts:5]: this should be a type import. But also it seems like this file owns this type definition not the other way round, so I'd suggest declaring and exporting it here, rather than the other way round

Comment (MrLoh) [reply src/window-dimensions.ts:5]: That also avoids the circular dependency which is always a sign of something odd going on

File: src/make-styles.tsx [hunk 3]
```diff
@@ -222,18 +250,19 @@ export const makeTemplateFunction = <
...
       const transformedProps = transformProps({ ...props, theme } as AttrProps<P, I, A>);
       const cssString = useMemo(() => {
         return resolveTemplateLiteral(strings, expressions, transformedProps);
       }, [transformedProps]);
       const styles = useMemo(() => createNestedStyleObject(cssString), [cssString]);
-      let styleProps: { [key: string]: Style | Style[] } = useStyleSheet(styles, theme, dimensions);
+      let styleProps: { [key: string]: Style | Style[] } = useStyleSheet(styles, theme, dimensions, containerDimensions[0]);
       styleProps = style ? { ...styleProps, style: [styleProps.style, style] } : styleProps;
       return (
         <Component {...filterComponentProps(transformedProps)} {...styleProps} ref={ref}>
```
Comment (MrLoh) [inline src/make-styles.tsx:268]: can't components that don't depend on props also be containers?

Comment (MrLoh) [reply src/make-styles.tsx:268]: I think this also needs to set the container context in other words

File: src/make-styles.tsx [hunk 4]
```diff
@@ -222,18 +250,19 @@ export const makeTemplateFunction = <
...
     StyledForwardRefRenderFunction = (
       // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
-      // @ts-ignore ts doesn't understand that A is not allwoed to declar children and style
+      // @ts-ignore ts doesn't understand that A is not allowed to declare children and style
       { children, style, ...props }: Omit<P & I, RequiredKeys<A>>,
       ref
     ) => {
       const theme = useTheme();
       const dimensions = useWindowDimensions();
+      const containerDimensions = useContainerDimensions();
```
[the remark this answers is not in the document]
Comment (MrLoh) [reply src/make-styles.tsx:259]: Also this doesn't seem to fallback to window dimensions

File: src/make-styles.tsx [hunk 5]
```diff
@@ -144,18 +155,19 @@ const matchMediaRule = (mediaRule: string, theme: Theme, windowDimensions: Scale
...
     // we need to make sure to do a deep clone here, so that theme and viewport units can be resolved from original strings
     const stylesCopy = { ...styles };
     for (const key in stylesCopy) {
       const { main, ...mediaStylesCopy } = stylesCopy[key];
-      // this will contain the main style and all applicapble media query styles
+      // this will contain the main style and all applicable media and container query styles
       const mediaStylesArray = [resolveThemeVariables({ ...main }, theme, windowDimensions)];
       for (const mediaRule in mediaStylesCopy) {
-        if (matchMediaRule(mediaRule, theme, windowDimensions)) {
+        if(matchQueryRule(mediaRule, theme, windowDimensions, containerDimensions.height > 0 ? containerDimensions : undefined)) {
```
Comment (MrLoh) [inline src/make-styles.tsx:170]: it seems to me like this will always use container dimensions no matter whether the rule is declared as `@media` or `@container` if a container context is present

File: src/make-styles.tsx [hunk 6]
```diff
@@ -100,11 +110,12 @@ export const createNestedStyleObject = (cssDeclaration: string): NestedStyles =>
       if (nOpen === 0) {
         const declaration = cssDeclaration.substring(start, match.index).trim();
         start = match.index + 1;
-        if (name!.substring(0, 6) === '@media') {
```
Comment (MrLoh) [inline src/make-styles.tsx:103]: why not just do `if(name!.substring(0, 6) === '@media')` and then `else if(name!.substring(0, 10) === '@container') ...` seems more readable to me. I'd always avoid regex if not really necessary they are extremely hard to read.

Comment (MrLoh) [reply src/make-styles.tsx:103]: and I think we need to preserve the info if this was `@media` or `@container` here

**Review by MrLoh — CHANGES_REQUESTED** [review]
This looks like it should work, very impressive. I think there are a few small bugs here though. I also think we could abstract the container stuff better, let's talk about that in person. But most importantly this is missing tests and we really should add tests especially with all those strange edge cases that might have some bugs hiding in them.

File: README.md [hunk 7]
```diff
@@ -51,6 +51,17 @@ const StyledComponentWithMediaQuery = styled.Text`
     font-size: 14px;
   }
 `;
+
```
Comment (MrLoh) [inline README.md:54]: we should mention container queries more prominently in the readme somewhere and provide a link to the spec

File: src/window-dimensions.ts [hunk 8]
```diff
@@ -1,19 +1,32 @@
...
-let timeoutId: NodeJS.Timeout | undefined;
-Dimensions.addEventListener('change', ({ window }) => {
-  timeoutId && clearTimeout(timeoutId);
-  timeoutId = setTimeout(() => {
-    initialWindowDimensions = window;
-    windowDimensionListeners.forEach((listener) => listener(window));
-  }, WINDOW_DIMENSIONS_DEBOUNCE);
-});
+
+const debounce = (fn: (...args: any[]) => void, duration = 100) => {
```
Comment (MrLoh) [inline src/window-dimensions.ts:11]: this might belong in a separate util and the `useContainerDimensions` doesn't have much to do with window dimensions, I wouldn't put all this code in one file probably because it has no semantic connection

**Comment by MrLoh** [comment]
I fixed the broken tests on master in 0cc2fdce2aff956ccbd775f4e580101b55b794a0

File: tests/container-queries.test.tsx [hunk 9]
```diff
@@ -0,0 +1,367 @@
+import { createNestedStyleObject } from '../src/make-styles';
+import React from 'react';
+import { ScrollView, Text } from 'react-native';
+
+import styled from '../src';
+import { render, sleep, theme } from './test-helper';
+
+describe('media queries', () => {
```
Comment (MrLoh) [inline tests/container-queries.test.tsx:8]: Should be Container queries

File: tests/container-queries.test.tsx [hunk 10]
```diff
@@ -0,0 +1,367 @@
...
+  beforeAll(() => jest.useFakeTimers());
+  afterAll(() => jest.useRealTimers());
+
+  beforeEach(() => {
+    (ScrollView as unknown as jest.Mock).mockClear();
```
Comment (MrLoh) [inline tests/container-queries.test.tsx:13]: Are we actually using any Scroll views?

File: tests/container-queries.test.tsx [hunk 11]
```diff
@@ -0,0 +1,367 @@
...
+    (Text as unknown as jest.Mock).mockClear();
+  });
+
+  const StyledFunctionContainer = styled.View<{ width: number; height?: number }>`
+    container: layout;
+    width: ${(p) => p.width}px;
+    height: ${(p) => (p.height ? p.height + 'px' : '20px')};
```
Comment (MrLoh) [inline tests/container-queries.test.tsx:20]: It's always write this as `${p => p.height || 20}px`

File: tests/container-queries.test.tsx [hunk 12]
```diff
@@ -0,0 +1,367 @@
...
+      {}
+    );
+  });
+
+  it('supports container and media queries', () => {
+    const StyledMediaComponent = styled.Text`
+      font-size: 2rem;
+      @media (max-width: 800px) {
+        font-size: 1.5rem;
+      }
```
Comment (MrLoh) [inline tests/container-queries.test.tsx:168]: I would expect this test case to have both container and media queries probably

File: tests/container-queries.test.tsx [hunk 13]
```diff
@@ -0,0 +1,367 @@
...
+    `;
+
+    //both cases render font size 1.5rem even with the container size changing
```
Comment (MrLoh) [inline tests/container-queries.test.tsx:171]: I would add a case here that tests that the media query still fires correctly with the same way that is mocked in the media query tests

File: tests/container-queries.test.tsx [hunk 14]
```diff
@@ -0,0 +1,367 @@
...
+    const StyledContainerWithName = styled.View`
+      container-name: sidebar;
+    `;
+    try {
+      render(<StyledContainerWithName />);
+      expect(false);
+    } 
+    catch(e) {
+      if((e as Error).message === "Container-name is not currently supported by styled-native-components") {
+        expect(true);
```
Comment (MrLoh) [inline tests/container-queries.test.tsx:353]: There might be an expect error. Not sure if that can be used here.

File: tests/container-queries.test.tsx [hunk 15]
```diff
@@ -0,0 +1,367 @@
...
+
+  const StyledContainerWithNameAndProps = styled.View<{active: boolean}>`
+      container-name: sidebar;
+      font-size: ${(p) => (p.active ? '2rem' : '1.5rem')};
+    `;
+    expect(() => {
+      render(<StyledContainerWithNameAndProps active/>);
+    }).toThrow('Container-name is not currently supported by styled-native-components');
+  });
+});
```
Comment (MrLoh) [inline tests/container-queries.test.tsx:367]: I think we are missing a test for a component both being a container and using a container query.

**Review by MrLoh — CHANGES_REQUESTED** [review]
Just a few details

**Comment by MrLoh** [comment]
@MorganWheel is this ready to merge?

**Comment by MorganWheel** [comment]
> @MorganWheel is this ready to merge?

Yup, I can merge it by the end of today unless I hear otherwise from you
